File: modules/kscan-pii-native/android/src/main/java/expo/modules/kscanpiinative/NativePrivacyConstants.kt
package expo.modules.kscanpiinative

object NativePrivacyConstants {
    const val MODULE_NAME = "KScanPiiNative"

    const val SANITIZER_VERSION = "native-face-mask-poc-1.0.0"
    const val DETECTOR_VERSION_MLKIT = "16.1.7"
    const val DETECTOR_VERSION_APPLE_VISION = "1"

    const val ACCEPTED_URI_SCHEME = "file"
    val ACCEPTED_MIME_TYPES = setOf("image/jpeg", "image/png")
    const val OUTPUT_MIME_TYPE = "image/png"
    const val OUTPUT_EXTENSION = "png"

    const val MAX_WIDTH = 4096
    const val MAX_HEIGHT = 4096
    const val MAX_PIXELS = 16_777_216L

    const val DEFAULT_PADDING_RATIO = 0.15
    const val MIN_PADDING_RATIO = 0.0
    const val MAX_PADDING_RATIO = 0.5

    const val IOU_DEDUPLICATION_THRESHOLD = 0.5

    // Opaque black.
    const val REDACTION_COLOR_R = 0
    const val REDACTION_COLOR_G = 0
    const val REDACTION_COLOR_B = 0
    const val REDACTION_COLOR_A = 255

    const val CHECKSUM_ALGORITHM = "fnv1a-dual-lane-64"

    const val CACHE_NAMESPACE = "kscan-pii-native"
    const val OUTPUT_FILE_PREFIX = "kscan-pii-"

    // Match TypeScript outward rounding policy from the audited POC.
    // Start edges are floored; end edges are ceiled.

    // ── License-plate region screening (Build 34, closet media privacy) ──────
    //
    // These live beside the face constants rather than in a second object
    // because plate masking is not a second capability. It is the same privacy
    // pipeline — same decoder, same normalizer, same redactor, same verifier,
    // same irreversible opaque rectangle — asked a different question about
    // WHICH regions to destroy. Only the selection differs, and the selection
    // is entirely described by the thresholds below. Nothing here is tuned
    // against a measured corpus; every value is a stated judgement that must be
    // checked on a physical build.

    const val PLATE_SANITIZER_VERSION = "native-plate-mask-1.0.0"

    /**
     * Bundled ML Kit Latin text recognition. The artifact carries the model in
     * the APK, so screening runs with no network and no first-run download —
     * the same reason the Play-delivered variant was rejected for the face and
     * pose detectors. See android/build.gradle for the full rationale.
     */
    const val DETECTOR_VERSION_MLKIT_TEXT = "16.0.1"
    const val PLATE_DETECTOR_IMPLEMENTATION = "mlkit_text_bundled"

    /**
     * ── THE PLATE GEOMETRY BAND ─────────────────────────────────────────────
     *
     * The detector receives TEXT REGIONS and must decide which of them are
     * plate-like WITHOUT reading a single character (see AndroidPlateDetector
     * for why that constraint is absolute). Shape is therefore the only
     * evidence available, and these five thresholds are the whole decision.
     *
     * Narrowest plate this band is expected to meet: a US/Canada passenger
     * plate is 12x6 inches — exactly 2:1 — so 2.0 is the floor and the
     * comparison is inclusive. Anything squarer than that is a word, a label,
     * a caption, a garment tag, a line of a book.
     *
     * KNOWN GAP, stated rather than hidden: motorcycle plates and the square
     * formats used in parts of Asia sit nearer 1.4:1 and are NOT selected by
     * this band. Lowering the floor to catch them would pull in most ordinary
     * short text in the frame, and a pipeline that blacks out every word on a
     * garment is one users switch off.
     */
    const val PLATE_MIN_ASPECT_RATIO = 2.0

    /**
     * A single-row EU plate (520x110mm) is roughly 4.7:1. ML Kit's box hugs the
     * CHARACTERS, which are inset from the plate's edges, so the measured ratio
     * of a real plate's text can run wider than the plate itself. 6.5 leaves
     * that headroom while still rejecting the long thin regions that dominate
     * street photography — banners, shop fascias, road signs, and any line of
     * running body text, all of which are far longer than this.
     */
    const val PLATE_MAX_ASPECT_RATIO = 6.5

    /**
     * Minimum region width as a fraction of image width.
     *
     * A plate-shaped region narrower than 3% of the frame is either far away or
     * incidental. Masking it stamps a visible black bar into the user's own
     * photo to hide a plate nobody could read from that photo. The trade is
     * deliberately biased toward fewer false positives; its cost is that a
     * plate on a car in the far background is not screened.
     */
    const val PLATE_MIN_WIDTH_RATIO = 0.03

    /**
     * Absolute minimum region height in pixels, independent of image size.
     *
     * Below roughly a dozen pixels of glyph height, plate characters are not
     * legible at capture resolution, so masking buys no privacy while the
     * false-positive cost stays real. The relative width test above cannot
     * express this on its own: on a small image, 3% of the width is a handful
     * of pixels. UNVALIDATED against real captures — verify on device.
     */
    const val PLATE_MIN_HEIGHT_PX = 12

    /**
     * Minimum region area as a fraction of image area (0.01%).
     *
     * A backstop, not a primary filter. It rejects the hairline region that
     * passes the width test but is one or two pixels tall in a large image —
     * geometry that satisfies every ratio above yet contains no readable
     * anything.
     */
    const val PLATE_MIN_AREA_RATIO = 0.0001

    // Padding deliberately REUSES the face constants (DEFAULT_PADDING_RATIO and
    // the MIN/MAX bounds). ML Kit's box wraps the glyphs, and the glyphs are
    // what identify the vehicle, so the same proportional margin that covers a
    // face box's slop covers a plate's. A caller that wants the plate's frame
    // and mounting covered too can raise paddingRatio within the same bounds.
}

File: modules/kscan-pii-native/ios/NativePrivacyConstants.swift
import Foundation

enum NativePrivacyConstants {
    static let moduleName = "KScanPiiNative"

    static let sanitizerVersion = "native-face-mask-poc-1.0.0"
    static let detectorVersionAppleVision = "1"

    static let acceptedUriScheme = "file"
    static let acceptedMimeTypes = Set(["image/jpeg", "image/png"])
    static let outputMimeType = "image/png"
    static let outputExtension = "png"

    static let maxWidth = 4096
    static let maxHeight = 4096
    static let maxPixels = 16_777_216

    static let defaultPaddingRatio = 0.15
    static let minPaddingRatio = 0.0
    static let maxPaddingRatio = 0.5

    static let iouDeduplicationThreshold = 0.5

    // Opaque black.
    static let redactionColorR: CGFloat = 0
    static let redactionColorG: CGFloat = 0
    static let redactionColorB: CGFloat = 0
    static let redactionColorA: CGFloat = 1

    static let checksumAlgorithm = "fnv1a-dual-lane-64"

    static let cacheNamespace = "kscan-pii-native"
    static let outputFilePrefix = "kscan-pii-"

    // ── License-plate REGION screening ──────────────────────────────────────
    //
    // A THIRD capability on this module, additive to face masking and person
    // detection and sharing their decoder, normalizer, redactor, verifier and
    // cache manager. It is deliberately NOT text recognition: the detector runs
    // VNDetectTextRectanglesRequest, which reports where text-shaped pixels are
    // and never what they say, so no plate number is ever produced in memory
    // let alone logged, returned or persisted. See IOSPlateDetector.
    //
    // Because the detector cannot read, the ONLY thing separating a plate from
    // any other text in the frame is the shape of its box. Every threshold
    // below is part of that one heuristic, so they live together here rather
    // than being spelled inline at the point each is compared.

    static let plateSanitizerVersion = "native-plate-mask-poc-1.0.0"
    static let plateDetectorImplementation = "vision_text_rectangles"
    /// Vision ships with the OS and exposes no separately versioned text model;
    /// this tracks the pinned request revision, not a downloaded asset.
    static let detectorVersionVisionTextRectangles = "1"

    /// Width:height band a text region must fall inside to be treated as
    /// plate-like.
    ///
    /// Lower bound 2.0: a US passenger plate is 12x6in — exactly 2.0 — and the
    /// glyph run Vision actually boxes inside it is narrower and therefore
    /// wider-looking still (roughly 4:1). 2.0 is the generous floor.
    ///
    /// Upper bound 6.5: European (520x110mm, 4.7:1) and Australian plates and
    /// their glyph runs top out near 6:1. 6.5 leaves headroom without admitting
    /// the long single lines of body text and signage that sit well above 8:1.
    ///
    /// KNOWN MISS, stated rather than hidden: stacked/motorcycle plates are
    /// roughly 7x4in (1.75:1) and fall BELOW this floor, so they are not
    /// screened. Lowering the floor to ~1.7 would cover them at the cost of
    /// masking substantially more non-plate text. That is a product call, not a
    /// code call, so the band is left where the contract specified it.
    static let plateMinAspectRatio = 2.0
    static let plateMaxAspectRatio = 6.5

    /// Minimum fraction of the image WIDTH the region must span. At the
    /// module's 4096px cap this is ~123px; on a 1024px image it is ~31px, which
    /// leaves a seven-character plate about 4px per glyph. Below this the region
    /// carries no legible plate at the resolutions this module accepts.
    ///
    /// This is the threshold most likely to be wrong in the field: "illegible to
    /// a person at 1x" is not the same claim as "unrecoverable", and a
    /// super-resolution model is a different adversary than an eye. Treat 0.03
    /// as a starting point to be tuned against real captures, not as a proof.
    static let plateMinRelativeWidth = 0.03

    /// Secondary guard, in fractions of total pixels. The width floor alone
    /// admits hairline strips — a region 5% of the frame wide and three pixels
    /// tall passes it — so area rejects what width cannot. Deliberately set
    /// BELOW the width floor's own implied area (a 4:1 region at 3% width on a
    /// 4:3 frame is ~0.0003) so that it never silently overrides the width rule
    /// it is meant to supplement.
    static let plateMinRelativeArea = 0.0002

    /// Absolute pixel floors, so a small input image cannot pass the relative
    /// tests with a region too few pixels across to hold a glyph row at all.
    static let plateMinPixelWidth = 24
    static let plateMinPixelHeight = 8

    /// There is deliberately NO upper size bound. An upper bound could only ever
    /// cause a genuine plate to be SKIPPED, and on a privacy boundary the safe
    /// direction of failure is to mask more, not less. The cost is that a very
    /// large wide text region — a sign, a banner, a wordmark across a garment —
    /// is masked as though it were a plate. That is a visible, recoverable
    /// product cost; the reverse is an unrecoverable privacy one.

    /// Plates get more padding than faces (0.15). Vision boxes the glyph RUN,
    /// not the plate, and frequently boxes only part of a run — a state name and
    /// a number can come back as two separate rectangles. The wider pad closes
    /// the border of the plate around the run, and makes the two rectangles of a
    /// split plate overlap enough for the existing IoU dedup to collapse them.
    static let defaultPlatePaddingRatio = 0.25
}
